include/seqan3/alignment/matrix/detail/traceback_path.hpp

```diff
@@ -0,0 +1,263 @@
+            if (static_cast<bool>(old_dir & trace_directions::left_open))
+                set_trace_direction(cur_dir); // Set the current trace_direction
+        }
+        else
+        {
+            assert(cur_dir == trace_directions::diagonal);
+
+            matrix_iter -= matrix_offset{matrix_dimension::horizontal};
+            matrix_iter -= matrix_offset{matrix_dimension::vertical};
+            set_trace_direction(cur_dir); // Set the current trace_direction
+        }
+        return *this;
+    }
+
+    //!\brief Returns an iterator advanced by one.
+    constexpr traceback_iterator operator++(int) noexcept
+    {
+        traceback_iterator tmp{*this};
+        ++(*this);
+        return tmp;
+    }
+    //!\}
+
+    /*!\name Comparison operators
+     * \{
+     */
+    //!\brief Returns `true` if both iterators are equal, `false` otherwise.
+    constexpr bool operator==(traceback_iterator const & rhs) const noexcept
+    {
+        return *matrix_iter == *rhs.matrix_iter;
+    }
+
+    //!\brief Returns `true` if the point-to-element is seqan3::detail::trace_directions::none.
+    constexpr bool operator==(std::ranges::default_sentinel_t const &) const noexcept
+    {
+        return *matrix_iter == trace_directions::none;
+    }
+
+    //!\brief copydoc operator==()
+    constexpr friend bool operator==(std::ranges::default_sentinel_t const &, traceback_iterator const & rhs) noexcept
+    {
+        return rhs == std::ranges::default_sentinel;
+    }
+
+    //!\brief Returns `true` if both iterators are not equal, `false` otherwise.
+    constexpr bool operator!=(traceback_iterator const & rhs) const noexcept
+    {
+        return !(*this == rhs);
+    }
+
+    //!\brief Returns `true` if the point-to-element is not seqan3::detail::trace_directions::none.
+    constexpr bool operator!=(std::ranges::default_sentinel_t const &) const noexcept
+    {
+        return !(*this == std::ranges::default_sentinel);
+    }
+
+    //!\brief copydoc operator!=()
+    constexpr friend bool operator!=(std::ranges::default_sentinel_t const &, traceback_iterator const & rhs) noexcept
+    {
+        return !(rhs == std::ranges::default_sentinel);
+    }
+    //!\}
+
+private:
+
+    /*!\brief Updates the current traceback direction.
+     * \param[in] dir The traceback direction to update.
+     */
+    void set_trace_direction(trace_directions & dir) const
+    {
+        if (static_cast<bool>(*matrix_iter & trace_directions::diagonal))
+            dir = trace_directions::diagonal;
+        else if (static_cast<bool>(*matrix_iter & trace_directions::up) ||
+                 static_cast<bool>(*matrix_iter & trace_directions::up_open))
+            dir = trace_directions::up;
+        else if (static_cast<bool>(*matrix_iter & trace_directions::left) ||
+                 static_cast<bool>(*matrix_iter & trace_directions::left_open))
+            dir = trace_directions::left;
+        else
+            dir = trace_directions::none;
```

Comment:
?

Reply:
Yes this is cur_dir. I've read about issues with private member functions, namely that it is hard to understand from the interface which variables are going to be modified. Accordingly the proposed solution is to mark private member functions const and pass the members that are modified as parameters to the function.